### PR TITLE
Fixes name resolution in Ubuntu 1804

### LINKED ISSUE #1151: No name resolution on new Ubuntu 18 vm
Apologies if this is not the correct place for this issue. 

My team is building templates with Packer and deploying VMs with `az group create deployment ...`, based loosely on these docs: https://docs.microsoft.com/en-us/azure/virtual-machines/linux/build-image-with-packer

I noticed that after the provisioning done by Packer, waagent deletes resolv.conf. This is fine, and works as expected when deploying a 16 vm. 

However as we attempt to move onto Ubuntu 18, when the vm is created, `/etc/resolv.conf` is missing, meaning it is not recreated when the vm is created from the disk image. `nslookup microsoft.com` times out, and no names are resolved when using eg curl. `systemd-resolve` however does work. 

Is this something waagent is responsible for?

### PR DESCRIPTION
**Pull request by hglkrijger**
- fixes #1151 

In prior versions of Ubuntu we remove `/etc/resolv.conf` on deprovision if it has been linked to anything but `/run/resolvconf/resolv.conf`. We have the option of keeping that pattern of behavior with 1804, and removing it if it no longer points to `/run/systemd/resolve/stub-resolv.conf`, however I believe the right fix is to change this behavior and stop deleting files we do not own.

### CODE REVIEW

File: azurelinuxagent/pa/deprovision/ubuntu.py [hunk 1]
```diff
@@ -40,3 +40,13 @@ def del_resolv(self, warnings, actions):
             files_to_del = ["/etc/resolvconf/resolv.conf.d/tail",
                             "/etc/resolvconf/resolv.conf.d/original"]
             actions.append(DeprovisionAction(fileutil.rm_files, files_to_del))
+
+
+class Ubuntu1804DeprovisionHandler(UbuntuDeprovisionHandler):
+    def __init__(self):
+        super(Ubuntu1804DeprovisionHandler, self).__init__()
+
+    def del_resolv(self, warnings, actions):
```
Comment (boumenot) [inline azurelinuxagent/pa/deprovision/ubuntu.py]: Is there any reason to keep the prior behavior?

**Comment by boumenot** [comment]
It looks like you have a lot of unnecessary commits in your history.

File: azurelinuxagent/pa/deprovision/ubuntu.py (hunk 1 above)
Comment (hglkrijger) [reply azurelinuxagent/pa/deprovision/ubuntu.py]: We still want the other actions implemented by the base handler.

Comment (boumenot) [reply azurelinuxagent/pa/deprovision/ubuntu.py]: Sorry for being unclear.  We are warning about something we used to do that was bad.  If this is bad we shouldn't be doing it now or in previous versions.  I believe this is the intent of #855.  It seems odd to warn about something you previously did illegally especially considering it *breaks* the VM.

You could also argue that we will address #855 in a later version as a single PR.

Comment (hglkrijger) [reply azurelinuxagent/pa/deprovision/ubuntu.py]: Changing established behavior in 1604 is highly risky, and I do not believe we should take the chance of breaking existing patterns. 1804 is a good opportunity to make that change though, and issuing a warning seems sensible if people have taken a dependency on that broken behavior, so they can change it. If you are arguing this should be an info message instead, sure.

#855 is specifically about cloud-init scripts, which is a sub-class of this problem. Aside from these two instances, I am not aware of any issues or asks in this area.

**Comment by Rangi-th** [comment]
The fix mentioned in #1164 is not found; as expected in the WA Linux Agent version 2.2.31

The file has only the below content - `/usr/lib/python3/dist-packages/azurelinuxagent/pa/deprovision/ubuntu.py`

```
import os
import azurelinuxagent.common.utils.fileutil as fileutil
from azurelinuxagent.pa.deprovision.default import DeprovisionHandler, \
    DeprovisionAction


class UbuntuDeprovisionHandler(DeprovisionHandler):
    def __init__(self):
        super(UbuntuDeprovisionHandler, self).__init__()

    def del_resolv(self, warnings, actions):
        if os.path.realpath(
                '/etc/resolv.conf') != '/run/resolvconf/resolv.conf':
            warnings.append("WARNING! /etc/resolv.conf will be deleted.")
            files_to_del = ["/etc/resolv.conf"]
            actions.append(DeprovisionAction(fileutil.rm_files, files_to_del))
        else:
            warnings.append("WARNING! /etc/resolvconf/resolv.conf.d/tail "
                            "and /etc/resolvconf/resolv.conf.d/original will "
                            "be deleted.")
            files_to_del = ["/etc/resolvconf/resolv.conf.d/tail",
                            "/etc/resolvconf/resolv.conf.d/original"]
            actions.append(DeprovisionAction(fileutil.rm_files, files_to_del))
```

**Pull request merged**